src/System.Xml.XPath/src/System/Xml/XPath/Internal/BooleanFunctions.cs

```diff
@@ -60,7 +60,10 @@ internal bool toBoolean(XPathNodeIterator nodeIterator)
         {
             object result = _arg.Evaluate(nodeIterator);
             if (result is XPathNodeIterator) return _arg.Advance() != null;
-            if (result is string) return toBoolean((string)result);
+
+            string str = result as string;
+            if (str != null) return toBoolean(str);
+
             if (result is double) return toBoolean((double)result);
             if (result is bool) return (bool)result;
```

Comment:
if (result is double) return toBoolean((double)result);
if (result is bool) return (bool)result;

->

double resultAsDouble = result as double;
if (resultAsDouble != null) return toBoolean(resultAsDouble);
bool resultAsBool = result as bool;
if (resultAsBool != null) return resultAsBool;

When I start to look at this code now I honestly start to think if previous readability isn't better than few mili/microseconds saved. Do you have any data on how faster it is in real scenario?